go sdk blog post (#2528)
pulumi/docs: content/blog/go-sdk-road-to-2/index.md

---

> title: "Modern Cloud Infrastructure in Golang  - The Road to 2.0"

Reviewer: ```suggestion
title: "Modern Cloud Infrastructure in Go  - The Road to 2.0"
```

authors: ["evan-boyle"]

> tags: ["go", "golang", "aws", "gcp", "azure"]

Reviewer: Do we need both a `go` and `golang` tag? (Our tags are getting busy!)

Reviewer 2: JFWIW - I opened https://github.com/pulumi/docs/issues/2532 just now on the general issue here.

> date: "2020-02-25"

Reviewer 3: ```suggestion
date: "2020-02-27"
```

meta_desc: "Pulumi + Go is a powerful combo for your cloud-native infrastructure."
meta_image: "pulumigo.png"
---

> Here at Pulumi, everyone on our engineering team is a Gopher. We build our open-source [pulumi/pulumi engine](https://github.com/pulumi/pulumi) and our SaaS with Go. Pulumi has supported Go in preview for more than a year, but we've recently invested in making Go support for Pulumi great as part of [Pulumi 2.0]({{< relref "/blog/pulumi-2-0-roadmap#better-language-support" >}}).

Reviewer: ```suggestion
Here at Pulumi, everyone on our engineering team is a Gopher. Go has quickly become the "language of the cloud," and so when we chose to build our open-source [pulumi/pulumi engine](https://github.com/pulumi/pulumi) and SaaS backend, we chose Go. As such, we are very excited to welcome Go to the family of supported infrastructure as code languages as part of [Pulumi 2.0]({{< relref "/blog/pulumi-2-0-roadmap#better-language-support" >}}).
```

## What is Pulumi?

Pulumi lets you use real languages to express your application’s infrastructure needs, using a powerful technique called “infrastructure as code.” Using infrastructure as code, you declare desired infrastructure, and an engine provisions it for you so that it’s automated, easy to replicate, and robust enough for demanding production requirements. Pulumi improves this process by leveraging real languages and making modern cloud infrastructure patterns, such as containers and serverless programs, first-class and easy.

> Automatically create, update, or delete cloud resources using Pulumi’s infrastructure as code engine, removing manual point-and-clicking in the cloud provider consoles, UIs, and ad-hoc scripts.

Reviewer: One bit of feedback in general is that it'd be great to say here why Go specifically is a great language for infrastructure as code. This reads as a very generic overview to Pulumi and it'd be nice to emphasize Go a little bit more FWIW.

Author: Updated.

> Automatically create, update, or delete cloud resources using Pulumi’s infrastructure as code engine, removing manual point-and-clicking in the cloud provider consoles, UIs, and ad-hoc scripts.

Reviewer 3: ```suggestion
- Automatically create, update, or delete cloud resources using Pulumi’s infrastructure as code engine, removing manual point-and-clicking in the cloud provider consoles, UIs, and ad-hoc scripts.
```

> Use your favorite IDEs and tools, including Visual Studio Code, taking advantage of features like auto-completion, refactoring, and interactive documentation.

Reviewer 3: ```suggestion
- Use your favorite IDEs and tools, including Visual Studio Code, taking advantage of features like auto-completion, refactoring, and interactive documentation.
```

> Catch mistakes early on with standard compiler errors, and an infrastructure-specific policy engine for enforcing security, compliance, and best practices.

Reviewer 3: ```suggestion
- Catch mistakes early on with standard compiler errors, and an infrastructure-specific policy engine for enforcing security, compliance, and best practices.
```

> Reuse any existing go packages, or distribute your own, whether that’s for infrastructure best practices, productivity, or just general programming patterns.

Reviewer 3: ```suggestion
- Reuse any existing go packages, or distribute your own, whether that’s for infrastructure best practices, productivity, or just general programming patterns.
```

> Build scalable cloud applications using classic infrastructure cloud-native technologies. including serverless functions, and highly scalable databases such as AWS Aurora, into your core development experience, bringing them closer to your application code.

Reviewer 3: ```suggestion
- Build scalable cloud applications using classic infrastructure cloud-native technologies. including serverless functions, and highly scalable databases such as AWS Aurora, into your core development experience, bringing them closer to your application code.
```

Pulumi’s free open source SDK, which includes a CLI and assortment of libraries, enables these capabilities.

> Creating an AWS S3 Bucket is familiar and simple:

Reviewer 3: ```suggestion
Creating an AWS S3 Bucket is a familiar and simple example that demonstrates these advantages:
```

```go
package main

import (
    "github.com/pulumi/pulumi-aws/sdk/go/aws/s3"
    "github.com/pulumi/pulumi/sdk/go/pulumi"
)

func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        // Create an AWS resource (S3 Bucket)
        bucket, err := s3.NewBucket(ctx, "my-bucket", nil)
        if err != nil {
            return err
        }

        // Export the name of the bucket
        ctx.Export("bucketName", bucket.ID())
        return nil
    })
}
```

> ## An Expressive and Idiomatic Type System

Reviewer: One aspect of the Go SDK that is worth mentioning is that we designed it to give you a little more control, with the intent of being embeddable within larger systems. That's why, for instance, you get the `*pulumi.Context` object. It's true we stopped short of programmatically driving deployments (per lots of recent discussions), however we've actually seen that play out in practice with some Go-oriented customers.

Author: This seems like a topic that could benefit from a post of its own similar to the type system deep dive. Much of this is undocumented and doesn't have public prior art to reference.

The goal of this post is to drive people through to the getting started and examples pages. Do you think that adding a section on programmatic access would help to this end, or would that topic be better as a standalone with some richer content?

Inputs, Outputs, and functions that operate over them ([All]({{< relref "/docs/intro/concepts/programming-model#all" >}}), [Apply]({{< relref "/docs/intro/concepts/programming-model#apply" >}}), [Sprintf]({{< relref "/docs/intro/concepts/programming-model#outputs-and-strings" >}})) must be compatible with raw type primitives (string, integer, list, map, boolean), and user-defined structs. In other words, we must support polymorphism in a language that doesn’t have generics. We hide the gritty details behind a rich, strongly-typed API that we were able to iterate on easily thanks to the utilization of [code generation](https://github.com/pulumi/pulumi/blob/master/sdk/go/pulumi/types_builtins.go).

## Try Pulumi + Go Today

> Although Pulumi for Go is in “preview” status, it supports all of the essential Pulumi programming model features (and the rest is on its way). Our goal is to gather feedback and over the next few weeks, and work hard to improve the Go experience across the board, including more examples and better documentation. To get started check out:

Reviewer 3: ```suggestion
Although Pulumi for Go is in “preview” status until 2.0 is released, it supports all of the essential Pulumi programming model features (and the rest is on its way). Our goal is to gather feedback and over the next few weeks, and work hard to improve the Go experience across the board, including more examples and better documentation. To get started check out:
```
